JUnit3 tests cannot be run with Gradle 9.3.0

### Current Behavior

```
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':lib:test'.
> Test process encountered an unexpected problem.
   > Received a completed event for test with unknown id '2.5'. Registered test ids: '[:lib:test, 2.1]'
```

### Expected Behavior

Expected tests to run and build to succeed.

### Context (optional)

_No response_

### Self-contained Reproducer Project

https://github.com/ulrikls/gradle-junit3

### Gradle version

9.3.0

### Gradle version that used to work

9.2.1

### Build scan URL (optional)

https://scans.gradle.com/s/pzu3n3mrvzf6g

### Your Environment (optional)

_No response_

--- a/platforms/jvm/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestEventAdapter.java
+++ b/platforms/jvm/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestEventAdapter.java
@@ -49,6 +49,7 @@ import java.util.ArrayDeque;
 import java.util.Arrays;
 import java.util.Deque;
 import java.util.HashMap;
+import java.util.Iterator;
 import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
 import java.util.List;
@@ -279,11 +280,24 @@ public class JUnitTestEventAdapter extends RunListener {
     }
 
     private void completeNode(TestNode node, TestCompleteEvent event) {
+        for (Iterator<TestNode> descendingIter = executingStack.descendingIterator(); descendingIter.hasNext();) {
+            TestNode current = descendingIter.next();
+            if (node.equals(current)) {
+                descendingIter.remove();
+                break;
+            }
+            // Complete any children that were not completed yet
+            if (node.equals(current.parent)) {
+                completeNode(current, new TestCompleteEvent(clock.getCurrentTime()));
+                // Re-initialize the iterator as the stack has changed
+                // This is expensive, but these cases should be rare, usually all children will have been completed already
+                descendingIter = executingStack.descendingIterator();
+            }
+        }
+        // The node is more likely to be at the end because we finish parents before children
+        getExecutingStackForCurrentThread().removeLastOccurrence(node);
         resultProcessor.completed(node.descriptor.getId(), event);
         executing.remove(node);
-        // The node is more likely to be at the end because we shouldn't finish parents before children
-        executingStack.removeLastOccurrence(node);
-        getExecutingStackForCurrentThread().removeLastOccurrence(node);
     }
 
     // Note: This is JUnit 4.13+ only, so it may not be called

--- a/platforms/jvm/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/junit4/AbstractJUnit4SuitesIntegrationTest.groovy
+++ b/platforms/jvm/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/junit4/AbstractJUnit4SuitesIntegrationTest.groovy
@@ -27,7 +27,12 @@ import static org.hamcrest.CoreMatchers.equalTo
 
 abstract class AbstractJUnit4SuitesIntegrationTest extends AbstractJUnitSuitesIntegrationTest {
     abstract boolean supportsSuiteOutput()
+
     abstract String getTestFrameworkJUnit3Dependencies()
+
+    boolean isVintage() {
+        return false
+    }
 
     @Override
     String getTestFrameworkSuiteImports() {
@@ -175,7 +180,7 @@ abstract class AbstractJUnit4SuitesIntegrationTest extends AbstractJUnitSuitesIn
         result.testPath(':org.gradle.ASuite:org.gradle.OtherTest:ok').onlyRoot().assertStderr(equalTo(""))
     }
 
-    def "supports Junit3 suites"() {
+    def "supports JUnit3 suites"() {
         given:
         file('src/test/java/org/gradle/SomeSuite.java') << """
             package org.gradle;
@@ -257,6 +262,78 @@ abstract class AbstractJUnit4SuitesIntegrationTest extends AbstractJUnitSuitesIn
             result.testPath("org.gradle.SomeSuite:org.gradle.SomeTest2").onlyRoot().assertStdout(containsString("stdout in TestSetup#teardown"))
             result.testPath("org.gradle.SomeSuite:org.gradle.SomeTest2").onlyRoot().assertStderr(containsString("stderr in TestSetup#teardown"))
         }
+    }
+
+    def "supports JUnit3 suites run by JUnit 4 suites"() {
+        given:
+        file('src/test/java/org/gradle/JUnit3Test.java') << """
+            package org.gradle;
+
+            import junit.framework.TestCase;
+            import junit.framework.TestSuite;
+            import org.junit.runner.RunWith;
+            import org.junit.runners.AllTests;
+
+            @RunWith(AllTests.class)
+            public class JUnit3Test extends TestCase {
+
+                public JUnit3Test(String name) {
+                    super(name);
+                }
+
+                public static TestSuite suite()
+                {
+                    var suite = new TestSuite("JUnit 3 style test suite");
+                    suite.addTest(new JUnit3Test("JUnit 3 style test") {
+                        @Override
+                        protected void runTest() {
+                            testSomeLibraryMethodReturnsTrue();
+                        }
+                    });
+                    return suite;
+                }
+
+                public void testSomeLibraryMethodReturnsTrue() {
+                    System.out.println("In JUnit 3 style test");
+                }
+            }
+        """.stripIndent()
+        file('src/test/java/org/gradle/JUnit4Suite.java') << """
+            package org.gradle;
+
+            import org.junit.runner.RunWith;
+            import org.junit.runners.Suite;
+
+            @RunWith(Suite.class)
+            @Suite.SuiteClasses(JUnit3Test.class)
+            public class JUnit4Suite {
+            }
+        """.stripIndent()
+        buildFile << """
+            apply plugin: 'java'
+            ${mavenCentralRepository()}
+            dependencies {
+                ${testFrameworkDependencies}
+            }
+            test {
+                ${configureTestFramework}
+                include '**/*Suite.class'
+                exclude '**/*Test.class'
+            }
+        """.stripIndent()
+
+        when:
+        executer.withTasks('test').run()
+
+        then:
+        GenericTestExecutionResult result = resultsFor()
+        def testNodeName = isVintage() ? 'org.gradle.JUnit3Test$1' : 'JUnit 3 style test'
+        result.assertTestPathsExecuted(
+            ':org.gradle.JUnit4Suite:JUnit 3 style test suite:' + testNodeName
+        )
+        result.testPath(':org.gradle.JUnit4Suite:JUnit 3 style test suite:' + testNodeName).onlyRoot()
+            .assertDisplayName(equalTo("JUnit 3 style test"))
+            .assertStdout(containsString("In JUnit 3 style test"))
     }
 
     def "can run multiple parameterized tests via suite only"() {

--- a/platforms/jvm/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/vintage/JUnitVintageSuitesIntegrationTest.groovy
+++ b/platforms/jvm/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/vintage/JUnitVintageSuitesIntegrationTest.groovy
@@ -39,4 +39,9 @@ class JUnitVintageSuitesIntegrationTest extends AbstractJUnit4SuitesIntegrationT
             testRuntimeOnly 'org.junit.platform:junit-platform-launcher'
         """
     }
+
+    @Override
+    boolean isVintage() {
+        return true
+    }
 }

